shell/browser/file_select_helper.cc

```diff
@@ -0,0 +1,299 @@
+#include "base/files/file_enumerator.h"
+#include "base/files/file_util.h"
+#include "base/strings/utf_string_conversions.h"
+#include "chrome/common/pref_names.h"
+#include "components/prefs/pref_service.h"
+#include "shell/browser/electron_browser_context.h"
+#include "shell/browser/javascript_environment.h"
+#include "shell/browser/ui/file_dialog.h"
+#include "shell/common/gin_converters/callback_converter.h"
+#include "shell/common/gin_converters/file_path_converter.h"
+
+using blink::mojom::FileChooserFileInfo;
+using blink::mojom::FileChooserFileInfoPtr;
+using blink::mojom::FileChooserParams;
+using blink::mojom::NativeFileInfo;
+
+namespace {
+void DeleteFiles(std::vector<base::FilePath> paths) {
+  for (auto& file_path : paths)
+    base::DeleteFile(file_path);
+}
+}  // namespace
+
+FileSelectHelper::FileSelectHelper(
+    content::RenderFrameHost* render_frame_host,
+    scoped_refptr<content::FileSelectListener> listener,
+    FileChooserParams::Mode mode)
+    : render_frame_host_(render_frame_host),
+      listener_(std::move(listener)),
+      mode_(mode) {
+  DCHECK(render_frame_host_);
+  DCHECK(listener_);
+
+  web_contents_ = content::WebContents::FromRenderFrameHost(render_frame_host);
+  DCHECK(web_contents_);
+  observer_.RemoveAll();
+  content::WebContentsObserver::Observe(web_contents_);
+  observer_.Add(render_frame_host_->GetRenderViewHost()->GetWidget());
+}
+
+FileSelectHelper::~FileSelectHelper() = default;
+
+void FileSelectHelper::ShowOpenDialog(
+    const file_dialog::DialogSettings& settings) {
+  v8::Isolate* isolate = electron::JavascriptEnvironment::GetIsolate();
+  v8::HandleScope scope(isolate);
+  gin_helper::Promise<gin_helper::Dictionary> promise(isolate);
+
+  // Will be released in one of the following
+  // OnFilesSelected - Callback flow completed with response
+  // RunFileChooserEnd - If user cancelled dialog or render_frame_host was
+  // destroyed WebContentsDestroyed
+  AddRefWithCheck();
+  DCHECK(HasAtLeastOneRef());
+
+  auto callback = base::BindOnce(&FileSelectHelper::OnOpenDialogDone,
+                                 base::Unretained(this));
+  ignore_result(promise.Then(std::move(callback)));
+
+  file_dialog::ShowOpenDialog(settings, std::move(promise));
+}
+
+void FileSelectHelper::ShowSaveDialog(
+    const file_dialog::DialogSettings& settings) {
+  v8::Isolate* isolate = electron::JavascriptEnvironment::GetIsolate();
+  v8::HandleScope scope(isolate);
+  gin_helper::Promise<gin_helper::Dictionary> promise(isolate);
+
+  // Will be released in one of the following
+  // OnFilesSelected - Callback flow completed with response
+  // RunFileChooserEnd - If user cancelled dialog or render_frame_host was
+  // destroyed WebContentsDestroyed
+  AddRefWithCheck();
+  DCHECK(HasAtLeastOneRef());
+
+  auto callback = base::BindOnce(&FileSelectHelper::OnSaveDialogDone,
+                                 base::Unretained(this));
+  ignore_result(promise.Then(std::move(callback)));
```

Comment:
It's very confusing that `file_dialog::Show{Open,Save}Dialog` call back to C++ code using a promise and a v8 dictionary, rather than a `base::OnceCallback` and plain C++ data.

Am I missing something about the reason the code is written that way? Can we refactor it to be more straightforward and avoid using v8 data structures to communicate between two C++ functions?

The relevant upstream code uses //ui/shell_dialogs, which exposes a reasonable interface for this that I think both the `dialog` API and this code could use.

Reply:
Yeah this mostly historic trying to use the exposed interface from `file_dialog` class and currently it only has the JS interfaces used for dialog apis.

> Can we refactor it to be more straightforward and avoid using v8 data structures to communicate between two C++ functions?

Yup we can and that actually is better. Thanks for identifying this!

> The relevant upstream code uses //ui/shell_dialogs, which exposes a reasonable interface for this that I think both the dialog API and this code could use.

This seems good to try, but maybe we do it in a follow-up PR, it would just increase the scope of the current PR.

Reply:
Planning on doing this in a refactor following this pr :)